sherpa/astro/utils/src/_utils.cc

```diff
@@ -338,25 +338,71 @@ namespace sherpa { namespace astro { namespace utils {
 			    &group ) )
       return NULL;
 
-    if( mask.get_size() == 0 ) {
-      PyErr_SetString( PyExc_TypeError,
-		       (char*)"mask array has no elements" );
+    if (mask.get_ndim() != 1) {
+      PyErr_SetString( PyExc_ValueError, (char*)"mask array must be 1D" );
+      return NULL;
+    }
+
+    if (group.get_ndim() != 1) {
+      PyErr_SetString( PyExc_ValueError, (char*)"group array must be 1D" );
+      return NULL;
+    }
+
+    // Use TypeError for the size=0 checks as this is what historically been
+    // used for this check (when only mask_size was checked).
+    //
+    npy_intp mask_size = mask.get_size();
+    npy_intp group_size = group.get_size();
+    if( mask_size == 0 ) {
+      PyErr_SetString( PyExc_TypeError, (char*)"mask array has no elements" );
+      return NULL;
+    }
+
+    if( group_size == 0 ) {
+      PyErr_SetString( PyExc_TypeError, (char*)"group array has no elements" );
+      return NULL;
+    }
+
+    if (group[0] < 0) {
+      PyErr_SetString( PyExc_ValueError, (char*)"The first element of group is negative" );
       return NULL;
     }
 
     if( EXIT_SUCCESS != res.zeros( group.get_ndim(), group.get_dims() ))
       return NULL;
 
-    npy_intp size = group.get_size();
     npy_intp jj = 0;
 
-    for( npy_intp ii = 0; ii < size; ++ii ) {
-      if( group[ ii ] >= 0 && ii != 0 )
+    // The group array starts a group with a value >= 0 and continues the
+    // group with a value < 0. The first element therefore starts a new
+    // group.
+    //
+    // The assumption is that the difference, in time, to just setting
+    //     res[ii] = mask[jj]
+    // is insignificant.
+    //
+    if( mask[jj] )
+      res[0] = true;
+
+    for( npy_intp ii = 1; ii < group_size; ++ii ) {
+      // Are we starting a new group?
+      //
+      if( group[ ii ] >= 0 ) {
```

Comment:
You can take advantage of compound conditional short circuiting here-
        if ((group[ii] >= 0) && (++jj >= mask_size) {
          PyErr_SetString( PyExc_ValueError, (char*) "More groups than mask elements" );
	  return NULL;
        }

Reply:
I personally find this much harder to read, as it relies on the `++jj` logic firing before the check. I know it's a standard C/C++ idiom, but I do feel like we are combining two things: the ++jj step which is meant to happen, and then a safety check.